### PR TITLE
Added missed API visibility annotations for public APIs and enable the check at the build time

### PR DESCRIPTION
**Pull request by reta**
### Description
Added missed API visibility annotations for public APIs and enable the check at the build time

### Related Issues
Closes https://github.com/opensearch-project/OpenSearch/issues/12871

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [X] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/build.gradle [hunk 1]
```diff
@@ -141,7 +141,7 @@ tasks.withType(JavaCompile).configureEach {
 
 compileJava {
   options.compilerArgs += ['-processor', ['org.apache.logging.log4j.core.config.plugins.processor.PluginProcessor',
-    'org.opensearch.common.annotation.processor.ApiAnnotationProcessor'].join(','), '-AcontinueOnFailingChecks']
+    'org.opensearch.common.annotation.processor.ApiAnnotationProcessor'].join(',')]
```
Comment (reta) [inline server/build.gradle:144]: The checks will become mandatory and will fail the build:

```
Note: Processing Log4j annotations                                                   
Note: Annotations processed                                                                                                                                                
Note: Processing OpenSearch Api annotations                                                                                                                                
error: The element org.opensearch.common.inject.ConfigurationException is part of the public APIs but is not marked as @PublicApi, @ExperimentalApi or @DeprecatedApi (referenced by org.opensearch.common.inject.spi.StaticInjectionRequest) 
Note: Processing Log4j annotations                                                                                                                                                                                                                                                                                                                    
Note: No elements to process                                                                                                                                               
Note: Processing OpenSearch Api annotations                                                                                                                                
Note: Some input files use or override a deprecated API.                                                                                                                   
Note: Recompile with -Xlint:deprecation for details.                                                                                                                       
1 error                                                                                                                                                                    
                                                                                                                                                                           
> Task :server:compileJava FAILED                  
```

**Comment by reta** [comment]
@peternied closing the loop here :-)

File: server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java [hunk 2]
```diff
@@ -52,7 +53,10 @@
 
 /**
  * An action that forwards REST requests to an extension
+ *
+ * @opensearch.api
  */
+@PublicApi(since = "2.7.0")
```
Comment (peternied) [inline server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java:59]: Seems like this should be marked as experimental

**Comment by peternied** [comment]
> If none of those classes can be changed on 2.x, our hands are going to be pretty tightly tied in terms of our ability to make changes.

@msfroh You mean no _breaking_ changes.  Is there a component that you think should be marked as experimental so we have that flexiblity?

**Comment by reta** [comment]
> I'm a little concerned about the number of APIs being marked as public, rather than internal (not that this PR is necessarily changing things -- it looks like it's just making explicit what was previously implicit). If none of those classes can be changed on 2.x, our hands are going to be pretty tightly tied in terms of our ability to make changes.

@msfroh all these APIs are exposed though OpenSearch plugins API (Plugin & Co) - so this not something we could change, this is just a fact, and it will immediately break if the modifications are not backward compatible.

**Comment by reta** [comment]
> I would lean toward maintaining backward compatibility of user API and data, and drop developer API compatibility within major versions, but that's probably a controversial take that extends beyond a single PR.

This out policy, this change ensures that at least what are exposed though OpenSearch plugins API (Plugin & Co) - is annotated. It won't break on API change at the moment (see please https://github.com/opensearch-project/OpenSearch/issues/9305) but it will break the build if the APIs are not annotated but leak publicly.

**Comment by reta** [comment]
> ❌ Gradle check result for [a7166fa](https://github.com/opensearch-project/OpenSearch/commit/a7166faca7de163ab4ac32cb8fcdec22c748c07b): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/35753/)

https://github.com/opensearch-project/OpenSearch/issues/12788

**Pull request merged**